sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,476 @@
+
+        res = self._run("org.freedesktop.DBus.Properties.Get", self.arg_interface, self.arg_name)
+        if res.rc != 0:
+            raise RuntimeError(res.stderr)
+
+        ret = self.type.clone()
+        ret.parse(DBUSResult(res.stdout))
+        return ret.value
+
+    def set_value(self, value) -> None:
+        """
+        Set the property's value
+        """
+        if not self.writable:
+            raise RuntimeError("Execution failed: not a writable property")
+
+        val = self.type.clone()
+        val.value = value
+
+        res = self._run("org.freedesktop.DBus.Properties.Set", self.arg_interface, self.arg_name, val)
+        if res.rc != 0:
+            raise RuntimeError(res.stderr)
+
+
+class ProxyMethod(ProxyObject):
+    """
+    A D-Bus method accessible as an object
+    """
+
+    def __init__(
+        self,
+        host: MultihostHost,
+        dest: str,
+        objpath: str,
+        bus: str,
+        /,
+        interface: str | None = None,
+        child: xml.Element | None = None,
+        input: list[DBUSType] | None = None,
+        output: list[DBUSType] | None = None,
+    ):
+        super().__init__(host, dest, objpath, bus, interface, child)
+        self.input = input
+        self.output = output
+
+    def __call__(self, *args) -> Any:
+        """
+        Execute the method
+        """
+        # In some internal cases we don't provide the child nor the interface at
+        # at instantiation because we don't use them, but on normal calls they are
+        # required.
+        if self.child is None:
+            raise RuntimeError("Execution failed: no child.")
+        if self.interface is None:
+            raise RuntimeError("Execution failed: no interface.")
+
+        objargs = []
+        if self.input is not None:
+            for i in range(len(self.input)):
+                obj = self.input[i].clone()
+                try:
+                    obj.value = args[i]
+                except IndexError:
+                    raise RuntimeError(
+                        f"Execution failed: {len(self.input)} " f"arguments required but only {i} provided."
+                    )
+                objargs.append(obj)
+
+        method = f'{self.interface}.{self.child.attrib["name"]}'
+        res = self._run(method, *tuple(objargs))
+        if res.rc != 0:
+            raise RuntimeError(f'Execution of \'{self.child.attrib["name"]}{args}\' failed: ' + res.stderr)
+
+        if self.output is None:
+            ret = None
+        else:
+            values = []
+            result = DBUSResult(res.stdout)
+            for i in range(len(self.output)):
+                obj = self.output[i].clone()
+                obj.parse(result)
+                values.append(obj.value)
+
+            ret = tuple(values) if len(values) > 1 else values[0]
+
+        return ret
+
+
+class DBUSObject(MultihostUtility[MultihostHost]):
```

Comment:
MultihostUtility should only be the top level class that you add to the role object.

Reply:
Done.